### PR TITLE
Add script to convert GGMLv3 LLaMA models to GGUF

### PR DESCRIPTION
**Pull request by KerfuffleV2**
Currently in a pretty reasonable state. Testing/feedback would be appreciated.

Converted file tested to parse these prompts to the same tokens as pre-GGUF llama.cpp:

1. `你喜欢小狗吗？`
2. `Once upon a time, in a dark forest, there lived a little fox`

I also tested these models with the second prompt:

1. Random LLaMA1 7B
2. `openorca-platypus2-13b.ggmlv3.q5_K_M.bin`
3. `gplatty-30b-superhot-8k.ggmlv3.q4_K_M.bin`
4. `platypus2-70b-instruct.ggmlv3.q4_K_M.bin`

Identical generation compared to loading the actual GGML file with pre-GGUF llama.cpp when specifying a seed.

*Note*: When testing, be sure to specify `--eps` and `--gqa` as is appropriate. You'll probably also want to specify `--context-length` (it defaults to `2048`).

*edit*: It's now possible to use HF or "original" format metadata like vocab when converting. Some information about this and the current state of the pull: https://github.com/ggerganov/llama.cpp/pull/2682#issuecomment-1685968053

Some perplexity results here: https://github.com/ggerganov/llama.cpp/pull/2682#issuecomment-1685885492

### CODE REVIEW

**Comment by Dampfinchen** [comment]
Nice. With the breaking change coming, such a script is crucial so many people can keep using their models!

**Comment by KerfuffleV2** [comment]
@TheBloke Would something like this actually be useful for you? It still requires rewriting the whole file, but should be a lot faster than converting from HF or `.pth` format and then quantizing.

One deficiency currently is that it has to try to revert the vocabulary mangling stuff the initial conversion to GGML performed and it doesn't seem possible to do this 100% correctly. However, I could potentially add a way to load the vocab from the original model metadata (`tokenizer.model`, `tokenizer_config.json`, `config.json`) and use that rather than the vocab in the GGML file.

**Comment by klosax** [comment]
You should add a warning that models converted without a a new copy of the needed vocab parts may not be fully functional in the future. There is on going work done on the tokenizer in llama.cpp and there could be issues later on without the additional data.

**Comment by klosax** [comment]
If you load the vocab from the original model during conversion, you could compare the model with a real gguf model using sha256sum to verify that your conversion script works.

**Comment by KerfuffleV2** [comment]
> You should add a warning that models converted without a a new copy of the needed vocab parts may not be fully functional in the future. 

There's already pretty big warning every time it runs:

    === WARNING === Be aware that this conversion script is best-effort. Use a native GGUF model if possible. === WARNING ===

Is that not enough?

I'm also not sure what you mean about "needed vocab parts". I don't think parts are necessarily missing, it's just special meta information like which tokens are "unknown" or whatever isn't really possible to recover.

> If you load the vocab from the original model during conversion, you could compare the model with a real gguf model using sha256sum to verify that your conversion script works.

Maybe that approach could work for just the vocab part (but there's probably an easier way).

I've been looking at the existing conversion scripts and it's not exactly clear what exactly the correct approach here is. For example, `convert.py` doesn't even add token types. `convert-llama-hf-to-gguf.py` does. Also kind of annoying how the latter just has everything at the top level so you can't import and reuse functions.

Anyway, assuming I did implement loading from `tokenizer.json` or whatever to build a new copy of the vocab rather than using  what was in the GGML file I'd just reuse existing code so the vocab part section at least would be exactly the same as the other conversion scripts. There's no reason to write a custom version of that.

**Comment by klosax** [comment]
> There's already pretty big warning every time it runs:

Ok good.

> I've been looking at the existing conversion scripts and it's not exactly clear what exactly the correct approach here is. For example, `convert.py` doesn't even add token types. `convert-llama-hf-to-gguf.py` does.

To test things out we made the simpler `convert-llama-hf-to-gguf.py` scripts first, since the `convert.py` is so complex. The latter is not fully finished yet and work is being done here https://github.com/ggerganov/llama.cpp/pull/2668. 

Beside a full copy of the vocab and scores we should have token types and the special token mapping for eos/bos etc.

File: convert-llama-ggmlv3-to-gguf.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                    n_kv_head = x
+            assert n_kv_head is not None, "Couldn't determine n_kv_head from GQA param"
+            print(f'- Guessed n_kv_head = {n_kv_head} based on GQA {cfg.gqa}')
+        self.n_kv_head = n_kv_head
+        self.name_map = gguf.get_tensor_name_map(gguf.MODEL_ARCH.LLAMA, ggml_model.hyperparameters.n_layer)
+
+    def save(self):
+        print('* Preparing to save GGUF file')
+        gguf_writer = gguf.GGUFWriter(self.cfg.output, gguf.MODEL_ARCH_NAMES[gguf.MODEL_ARCH.LLAMA], use_temp_file = False)
+        self.add_params(gguf_writer)
```
Comment (Green-Sky) [inline convert-llama-ggmlv3-to-gguf.py:152]: maybe add a `gguf_writer.add_description('converted from old ggjtv3 file.')`
or something.

**Comment by KerfuffleV2** [comment]
> Beside a full copy of the vocab and scores we should have token types and the special token mapping for eos/bos etc.

Seems reasonable, and once `convert.py` is updated I can just use that to load the vocab and my GGUF output for vocab at least should be exactly the same as the official conversion.

I don't want to go too crazy with the amount of work I put into this when so far there's on indication that it's a candidate to get merged. Also, converting from GGML but also requiring the HF metadata seems like it would be kind of a niche use and no one's said "I'd actually use this feature!" yet.

File: convert-llama-ggmlv3-to-gguf.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            # print(f'+ {tensor.name} | {mapped_name} {tensor.dims} :: {tempdims}')
+            gguf_writer.add_tensor(mapped_name, data[tensor.start_offset:tensor.start_offset + tensor.len_bytes], raw_shape = tempdims, raw_dtype = tensor.dtype)
+
+
+def handle_args():
+    parser = argparse.ArgumentParser(description = 'Convert GGMLv3 models to GGUF')
+    parser.add_argument('--input', '-i', help = 'Input GGMLv3 filename')
+    parser.add_argument('--output', '-o', help ='Output GGUF filename')
+    parser.add_argument('--gqa', type = int, default = 1, help = 'grouped-query attention factor (use 8 for LLaMA2 70B)')
+    parser.add_argument('--eps', default = '5.0e-06', help = 'RMS norm eps (use 1e-5 for LLaMA2)')
```
Comment (klosax) [inline convert-llama-ggmlv3-to-gguf.py:230]: For LLaMA v1 / OpenLLaMA models the epsilon is 1e-6. The 5e-6 was a compromise of those two.

**Comment by Green-Sky** [comment]
> seems like it would be kind of a niche use 

not everyone has endless high speed internet access :)

File: convert-llama-ggmlv3-to-gguf.py (hunk 1 above)
Comment (KerfuffleV2) [reply convert-llama-ggmlv3-to-gguf.py:152]: Seems very reasonable.

File: convert-llama-ggmlv3-to-gguf.py (hunk 2 above)
Comment (KerfuffleV2) [reply convert-llama-ggmlv3-to-gguf.py:230]: I guess we'd still want to use that as the default value then but mention `1e-6` in the help?

Comment (KerfuffleV2) [reply convert-llama-ggmlv3-to-gguf.py]: I updated the help for `--eps` to be more specific about the values for different model types.

**Comment by KerfuffleV2** [comment]
@Green-Sky 

> not everyone has endless high speed internet access :)

So you're saying you need and would use this feature? If so, I'll look into adding it.

Probably need to wait until the `convert.py` vocab stuff stabilizes, hopefully that will happen at a point that gives me enough time to update this.

**Comment by klosax** [comment]
You should add a parameter to set the model name, and maybe default it to the filename. `gguf_writer.add_name()`

**Comment by TheBloke** [comment]
@KerfuffleV2 Thanks very much for working on this!

If it's confirmed that this will produce 100% identical results to a new convert.py conversion to GGUF then yes I would definitely use it, and save ~10+TB of data download :)

And if that is achieved, I think it absolutely should be merged.  There's many users out there on slow or metered internet connections who aren't going to relish re-downloading all their favourite GGML models. 

Great work!

**Comment by KerfuffleV2** [comment]
> If it's confirmed that this will produce 100% identical results to a new convert.py conversion to GGUF

I can understand why you'd say that but a guarantee like that isn't really possible. I'll do the best I can to ensure correct output, but I can't promise it's 100%. No one can even promise `convert.py` is going to always be correct either.

Even in cases where it's correct, output from a GGML conversion probably won't be _exactly_ the same as what convert.py would output because the order of tensors in the GGML file may differ and that kind of thing. That's not something that should actually cause a noticeable difference, but it still wouldn't be exactly the same file.

**Comment by KerfuffleV2** [comment]
I added the capability to override the hyperparameters and vocab from the HF or PyTorch metadata. Note that it leverages the existing `convert.py` script and right now that doesn't fully handle vocabulary types (it might actually be worse than converting from GGML).

For testing this, you can try copying in `convert.py` from #2668 - my conversion script should be able to handle either version of `convert.py` (assuming that pull doesn't add new changes that are incompatible).

**Comment by klosax** [comment]
I guess you dontt have to bother with other vocab types than the LLaMA spm. The chinese Aquila is the only LLaMA model using the gpt2 bpe tokenizer I know of. But if you are going to include that you will also need to include the merges from the original model, see the gptneox / falcon conversion examples in the gguf branch.

**Comment by KerfuffleV2** [comment]
> I'll definitely be using this conversion script as I have pretty slow internet myself. 

Please test it out and provide feedback if you're able to!

>  The chinese Aquila is the only LLaMA model using the gpt2 bpe tokenizer I know of. But if you are going to include that you will also need to include the merges from the original model, see the gptneox / falcon conversion examples in the gguf branch.

I'm just using `convert.py`'s vocabulary loader and there's a commandline option to specify vocab type also same as with `convert.py`. So it should just work, I think (as long as `convert.py` works).

**Comment by TheBloke** [comment]
> > If it's confirmed that this will produce 100% identical results to a new convert.py conversion to GGUF
> 
> I can understand why you'd say that but a guarantee like that isn't really possible. I'll do the best I can to ensure correct output, but I can't promise it's 100%. No one can even promise `convert.py` is going to always be correct either.
> 
> Even in cases where it's correct, output from a GGML conversion probably won't be _exactly_ the same as what convert.py would output because the order of tensors in the GGML file may differ and that kind of thing. That's not something that should actually cause a noticeable difference, but it still wouldn't be exactly the same file.

Yeah I quite understand. I just have to be cautious - the worst case scenario for me is where I convert thousands of files with the script, they all seem to work fine, and then 48 hours later there's a reports of issues in niche cases I'd not tested and I feel I have to do them all again.  It's unlikely to happen, I just have to weigh up the possibility when deciding what method to use.

Let me know if some HW would help your testing. Eg I could provide you a 4090 pod with fast CPU if that'd help you run perplexity tests, or speed up any other testing.

**Comment by KerfuffleV2** [comment]
@TheBloke 

>  I just have to weigh up the possibility when deciding what method to use.

Absolutely. Unfortunately, there are risks with both approaches. As far as I know, I'm the only one who's even tried this pull and having only one person test something is never going to be ideal. At least the official approach has a lot more eyes on it and more people trying it on various models, etc.

> Eg I could provide you a 4090 pod with fast CPU and 1 Gb/s internet if that'd help you run perplexity tests, or speed up any other testing.

That's a very generous offer. This would definitely help with perplexity tests _but_ I'm not 100% convinced perplexity is going to highlight the kind of issues that would occur from vocab conversion issues (which I think is the most likely problem).

Not sure if you have the time/inclination but it actually might be better for you to do some of those tests yourself. Since you're not me, you're likely to do stuff I wouldn't do and potentially run into issues that I wouldn't. Of course it's possible to use both approaches.

***

Let me also give you (and anyone else interested) a quick summary of the current state:

Converting without the external metadata (`config.json`, `tokenizer.model`, etc) from the original model might be good enough for general use but I'd say it's probably _not_ suitable for conversion/wide scale distribution like you're aiming for. Some information just isn't in the GGML file and can't be recovered without external stuff.

Converting with the external metadata _should_ produce a GGUF with the same parameters and vocab as you'd get converting from HF to GGUF directly. It's just the tensor data/info that is getting copied from the GGML file in that case and as far as I know nothing changed there between GGUF and GGML. The file might not be exactly identical (due to stuff like the order of the tensors) but it should be functionally the same.

**Note**: Using the external metadata just uses the existing `convert.py` and the one in the `gguf` branch lacks some features. This would be an issue converting from HF also but I'd recommend you use the `convert.py` from #2668 - you can just copy it into the directory this pull is checked out to. (But to make life more complicated, I don't think the version in #2668 has the fixed LLaMA2 70B permute stuff so don't use `convert.py` from that branch to convert HF models.)

Example of running conversion with external metadata:

    python convert-llama-ggmlv3-to-gguf.py --in blah.bin --out blah.gguf -m /path/to/metadata

It's also possible to specify `--vocab-dir` and `--vocabtype` same as the official conversion script. The metadata directory can be HF format or the "original" format. However, when using "original" format metadata if `n_vocab` or `n_ff` is missing from `params.json` it'll fail (it'll just crash, not do anything like produce incorrect output).

**Comment by klosax** [comment]
There is also a slight difference in perplexity between master and gguf with models converted from HF models.
(wiki.test.raw.406)
 model | master | gguf
 :-- | --: | --:
openllama-7b-v2-f16 | 7.16735811 | 7.16541492
openllama-7b-v2-q8_0 | 7.17201356| 7.17004949
openllama-7b-v2-q4_0 | 7.32185657 | 7.31998091

**Comment by Dampfinchen** [comment]
So basically for best results one needs to download the metadata of the FP16 model file from HF (tokenizer_config.json, config.json etc etc), put it in a folder and then use -m to direct the program to this folder? 

So for example for MythoMax... https://huggingface.co/Gryphe/MythoMax-L2-13b/tree/main you would download everything aside from the model  itself, and put that in a folder called Metadata. Since I use q4k_m, the command would look like this:

python convert-llama-ggmlv3-to-gguf.py --in "mythomax-l2-13b.ggmlv3.q4_K_M.bin" --out "mythomax-l2-13b.ggmlv3.q4_K_M.gguf" -m Metadata 

(Assuming the script is running in the same folder)

If I'm understanding it correctly, that's a pretty easy process to do.

**Comment by KerfuffleV2** [comment]
> So basically for best results one needs to download the metadata of the FP16 model file from HF (tokenizer_config.json, config.json etc etc), put it in a folder and then use -m to direct the program to this folder?

Since the model data won't be used, it doesn't have to be FP16 or anything in particular, but basically yes. If you have git lfs set up you can just do something like:

    env GIT_LFS_SKIP_SMUDGE=1 git lfs clone https://huggingface.co/user/model
    cd model
    git lfs pull --include=tokenizer.model

The smudge stuff tells git to only fetch pointers to the large files (so that'll skip all the model datafiles). You do need to actually fetch `tokenizer.model` though.

> If I'm understanding it correctly, that's a pretty easy process to do.

Yep, you understand correctly. Just note the part about using `convert.py` from #2668 (at least until it gets merged). If you use `convert.py` from this branch or `gguf` it'll potentially be worse than just trying to convert the vocab from the GGML file.

**Review by ggerganov — APPROVED** [review]
When you are ready, merge this to `gguf` 
I'll merge `gguf` to `master` in an hour or two. Alternatively, you can change the target branch to `master` and merge it after #2398

**Comment by KerfuffleV2** [comment]
@ggerganov 

> When you are ready, merge this to gguf

Sounds good. I think I'm done making changes unless other people request stuff or find issues. Do you have a preference over doing it now or waiting? This pull really benefits from #2668 but conversion in general also would. Hopefully that one can make it in also (merging or not won't require changes to this pull).

Also just want to double check that you saw the changes to `gguf.py` -

1. I added the ability to avoid using a temp file (my conversion stuff just uses numpy's memmap functionality so it doesn't have to explicitly get read into memory). I think the normal `convert.py` stuff could use that approach too but I didn't want to mess with existing stuff too much.
2. I made it possible to override the shape and tensor type for `add_tensor` and `add_tensor_info`.

Both of those changes are opt-in and the default behavior should be the same as without this pull.

**Comment by ggerganov** [comment]
> I think the normal convert.py stuff could use that approach too but I didn't want to mess with existing stuff too much.

Sure - improvements are welcome.

Will try to get #2668 merged in `gguf` as well before the final merge to `master`
